django/db/backends/sqlite3/base.py

```diff
@@ -266,37 +267,68 @@ def check_constraints(self, table_names=None):
         determine if rows with invalid references were entered while constraint
         checks were off.
         """
-        with self.cursor() as cursor:
-            if table_names is None:
-                table_names = self.introspection.table_names(cursor)
-            for table_name in table_names:
-                primary_key_column_name = self.introspection.get_primary_key_column(cursor, table_name)
-                if not primary_key_column_name:
-                    continue
-                key_columns = self.introspection.get_key_columns(cursor, table_name)
-                for column_name, referenced_table_name, referenced_column_name in key_columns:
-                    cursor.execute(
-                        """
-                        SELECT REFERRING.`%s`, REFERRING.`%s` FROM `%s` as REFERRING
-                        LEFT JOIN `%s` as REFERRED
-                        ON (REFERRING.`%s` = REFERRED.`%s`)
-                        WHERE REFERRING.`%s` IS NOT NULL AND REFERRED.`%s` IS NULL
-                        """
-                        % (
-                            primary_key_column_name, column_name, table_name,
-                            referenced_table_name, column_name, referenced_column_name,
-                            column_name, referenced_column_name,
+        if Database.sqlite_version_info >= (3, 20, 0):
+            with self.cursor() as cursor:
+                if table_names is None:
+                    violations = self.cursor().execute('PRAGMA foreign_key_check').fetchall()
+                else:
+                    violations = chain.from_iterable(
+                        cursor.execute('PRAGMA foreign_key_check(%s)' % table_name).fetchall()
+                        for table_name in table_names
+                    )
+                # See https://www.sqlite.org/pragma.html#pragma_foreign_key_check
+                for table_name, rowid, referenced_table_name, foreign_key_index in violations:
+                    foreign_key = cursor.execute(
+                        'PRAGMA foreign_key_list(%s)' % table_name
+                    ).fetchall()[foreign_key_index]
+                    column_name, referenced_column_name = foreign_key[3:5]
+                    primary_key_column_name = self.introspection.get_primary_key_column(cursor, table_name)
+                    primary_key_value, bad_value = cursor.execute(
+                        'SELECT %s, %s FROM %s WHERE rowid = %%s' % (
+                            primary_key_column_name, column_name, table_name
```

Comment:
Not worth amending now, but include a trailing comma in cases like this so that if more items are added later, this line doesn't have to be modified again. https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas